File: async/src/main/java/cai/lib/async/AsyncThreadFactory.java
package cai.lib.async;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

import android.os.Process;

/**
 * 线程创建工厂
 */
public class AsyncThreadFactory implements ThreadFactory {

    private static final AtomicInteger poolNumber = new AtomicInteger(1);
    private final AtomicInteger threadNumber = new AtomicInteger(1);
    private final ThreadGroup group;
    private final String namePrefix;

    public AsyncThreadFactory(String threadPoolName) {
        SecurityManager security = System.getSecurityManager();
        group = (security != null) ? security.getThreadGroup() : Thread.currentThread().getThreadGroup();
        namePrefix = threadPoolName + '-' + poolNumber.getAndIncrement() + '-';
    }


    @Override
    public Thread newThread(Runnable runnable) {
        runnable = runnable != null ? runnable : () -> { /* ignore */ };
        PriorityRunnable pr = new PriorityRunnable(runnable, Process.THREAD_PRIORITY_BACKGROUND);
        Thread thread = new Thread(group, pr, namePrefix + threadNumber.getAndIncrement(), 0);
        if (thread.isDaemon()) {
            thread.setDaemon(false);
        }
        if (thread.getPriority() != Thread.NORM_PRIORITY) {
            thread.setPriority(Thread.NORM_PRIORITY);
        }
        return thread;
    }
}

File: async/src/main/java/cai/lib/async/AsyncThreadFactory.kt
package cai.lib.async

import android.os.Process
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger

/**
 * 线程创建工厂
 */
class AsyncThreadFactory(threadPoolName: String) : ThreadFactory {

    companion object {
        private val poolNumber = AtomicInteger(1)
    }

    private val threadNumber = AtomicInteger(1)
    private val group: ThreadGroup
    private val namePrefix: String

    init {
        val security = System.getSecurityManager()
        group = if (security != null) security.threadGroup else Thread.currentThread().threadGroup!!
        namePrefix = threadPoolName + '-' + poolNumber.getAndIncrement() + '-'
    }

    override fun newThread(runnable: Runnable?): Thread {
        val run = runnable ?: Runnable {}
        val priorityRunnable = PriorityRunnable(run, Process.THREAD_PRIORITY_BACKGROUND)
        val thread = Thread(group, priorityRunnable, namePrefix + threadNumber.getAndIncrement(), 0)
        if (thread.isDaemon) {
            thread.isDaemon = false
        }
        if (thread.priority != Thread.NORM_PRIORITY) {
            thread.priority = Thread.NORM_PRIORITY
        }
        return thread
    }
}